### PR TITLE
Correct matrix assembler part

### PR DESCRIPTION
**Pull request by rewu1993**

### CODE REVIEW

File: include/matrix_assembler.h [hunk 1]
```diff
@@ -33,6 +33,8 @@ class MatrixAssembler {
   //! Obtain global nodal and pipe indices and pointers from meshes
   //! \param[in] mesh meshes that form the pipe network
   void global_nodal_pipe_indices(const std::shared_ptr<Mesh>& mesh);
+  void global_nodal_pipe_indices(const std::shared_ptr<pipenetwork::Mesh>& mesh,
+                                 const unsigned nnode_known);
```
Comment (kks32) [inline include/matrix_assembler.h]: No need to const this

File: CMakeLists.txt [hunk 2]
```diff
@@ -68,13 +69,16 @@ SET(test_src
...
    src/pipe.cc
    src/io.cc
-   tests/eigen_gmres_test.cc
-   tests/matrix_assembler_test.cc
-   tests/mesh_test.cc
-   tests/node_test.cc
-   tests/pipe_test.cc
-   tests/io_test.cc
+        src/node.cc
+#   tests/eigen_gmres_test.cc
```
Comment (kks32) [inline CMakeLists.txt]: Could you uncomment these?

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/matrix_assembler.h [hunk 3]
```diff
@@ -140,6 +92,25 @@ class MatrixAssembler {
...
   //! id map that maps node id to variable id
   std::shared_ptr<std::map<Index, Index>> id_map_;
+
+  //! Assemble pressure demand part of the jacobian matrix
+  //! \param[in] n the corresponding node pointer
+  //! \param[in] index the position of the corresponding node in variable vector
+  //! \param[in] update the vector that stores the position of element in
+  //! jacobian matrix
+  void construct_demand_jac(const std::shared_ptr<pipenetwork::Node> & n, Index index,
+                            std::vector<Eigen::Triplet<double>>& update);
```
Comment (kks32) [inline include/matrix_assembler.h]: Are we modifying this vector?

File: include/node.h [hunk 4]
```diff
@@ -42,73 +42,63 @@ class Node {
...
 
   //! Return hydraulic head
   //! \retval head_ hydraulic head at the node
+  double elevation() const { return elevation_; }
   double head() const { return head_; }
-  double iter_head() const { return iter_head_; }
 
   //! Return head assignment status
   //! \retval ishead_ head assignment status at the node
   bool ishead() const { return ishead_; }
```
Comment (kks32) [inline include/node.h]: Does this mean we can have both a head and a reservoir?

File: include/matrix_assembler.h [hunk 5]
```diff
@@ -70,34 +71,46 @@ class MatrixAssembler {
...
-
  private:
   //! global nodal id and corresponding nodal pointer
   std::map<Index, std::shared_ptr<pipenetwork::Node>> global_nodes_;
   //! global pipe id and corresponding pipe pointer
   std::map<Index, std::shared_ptr<pipenetwork::Pipe>> global_pipes_;
   //! number of nodes in the network
   unsigned nnode_{0};
+  //! number of known nodes in the network
+  unsigned nnode_known_{0};
```
Comment (kks32) [inline include/matrix_assembler.h]: Could we check this isn't > nnode_ and maybe some other value so we can always solve the equation

File: include/node.h [hunk 6]
```diff
@@ -42,34 +42,64 @@ class Node {
...
   //! \retval discharge_ discharge at the node
-  double discharge() const { return discharge_; }
+  double demand() const { return demand_; }
 
   //! Return discharge assignment status
   //! \retval isdischarge_ discharge assignment status at the node
   bool isdischarge() const { return isdischarge_; }
 
+  //! min pressure
+  void min_pressure(double min_pressure) { minimum_pressure_ = min_pressure; }
```
Comment (kks32) [inline include/node.h]: Do we need all these getters and setters?

File: src/eigen_gmres.cc [hunk 7]
```diff
@@ -30,7 +30,8 @@ bool pipenetwork::EigenGMRES::solve() {
   gmres.setTolerance(tolerance_);
 
   x_diff = gmres.solve(*vec_b_);
-  x_diff = x_diff.array() * vrestraints_.array();
+  x_diff = x_diff.array();
+  //    x_diff = x_diff.array() * vrestraints_.array();
```
Comment (kks32) [inline src/eigen_gmres.cc:34]: So we no longer need to apply any restraints?

File: src/matrix_assembler.cc [hunk 8]
```diff
@@ -1,10 +1,14 @@
-#include "matrix_assembler.h"
+
+#include <cmath>
```
Comment (kks32) [inline src/matrix_assembler.cc]: Move this to header

File: src/matrix_assembler.cc [hunk 9]
```diff
@@ -1,10 +1,14 @@
...
+#include <matrix_assembler.h>
```
Comment (kks32) [inline src/matrix_assembler.cc]: Should be "matrix_assembler.h"

File: src/matrix_assembler.cc [hunk 10]
```diff
@@ -31,111 +35,55 @@ void pipenetwork::MatrixAssembler::assemble_variable_vector() {
...
-  for (const auto& node : global_nodes_) {
-    Index index = node.first;
-    // Calculate the nodal balance residual (demand part)
-    if (node.second->isdischarge())
-      residual_vec_->coeffRef(index) -= (-1) * node.second->discharge();
+// assember the pressure demand part of jacobian matrix
+void pipenetwork::MatrixAssembler::construct_demand_jac(
+    const std::shared_ptr<pipenetwork::Node> & node, Index index,
+    std::vector<Eigen::Triplet<double>>& update) {
+  if (not node->isres()) {
```
Comment (kks32) [inline src/matrix_assembler.cc]: You can use `!` instead of not

File: src/matrix_assembler.cc [hunk 11]
```diff
@@ -194,122 +154,182 @@ void pipenetwork::MatrixAssembler::assemble_jacobian() {
...
+      residual_vec_->coeffRef(index) -= (-1) * node.second->iter_demand ();
+    // Calculate the demand pressure residual
+    // Case 1 the node is Reservior/tank
+    if (node.second->isres()) {
+      residual_vec_->coeffRef(nnode_ + npipe_ + index) =
+          node.second->head () - node.second->elevation ();
+    }
+    // case 2 the node is a junction
+    else {
+      if (not pdd_) {
```
Comment (kks32) [inline src/matrix_assembler.cc]: if (!pdd_)

File: src/matrix_assembler.cc [hunk 12]
```diff
@@ -194,122 +154,182 @@ void pipenetwork::MatrixAssembler::assemble_jacobian() {
...
-// equation, Thus the Jacobian has (nnode+npipe) row and (nnode+npipe) column
-//
-void pipenetwork::MatrixAssembler::sim_assemble_jacobian() {
-  // Check network
-  if (nnode_ <= 0 || npipe_ <= 0)
-    throw std::runtime_error(
-        "No node or pipe index pairs created to assemble Jacobian matrix");
+void pipenetwork::MatrixAssembler::assemble_pdd_residual(
+    const std::shared_ptr<pipenetwork::Node> & node, Index index) {
+  auto pressure = node->head () - node->elevation ();
```
Comment (kks32) [inline src/matrix_assembler.cc]:
```suggestion
  auto pressure = node->head() - node->elevation();
```

File: src/matrix_assembler.cc [hunk 13]
```diff
@@ -194,122 +154,182 @@ void pipenetwork::MatrixAssembler::assemble_jacobian() {
...
+  double res;
 
-  std::vector<Eigen::Triplet<double>> update;
-  update.reserve(5 * npipe_);
+  // case 1
```
Comment (kks32) [inline src/matrix_assembler.cc]: Instead of `case` can you comment what the case is?

File: src/matrix_assembler.cc [hunk 14]
```diff
@@ -194,122 +154,182 @@ void pipenetwork::MatrixAssembler::assemble_jacobian() {
...
+  if (pressure < node->min_pressure()) {
+    res = -1 * (node->iter_demand () - node->demand () * node->pdd_slope() *
+                                             (pressure - node->min_pressure()));
+  }
+  // case 2
+  else if ((pressure > node->min_pressure()) and
```
Comment (kks32) [inline src/matrix_assembler.cc]: instead of `and` use `&&`

File: src/matrix_assembler.cc [hunk 15]
```diff
@@ -194,122 +154,182 @@ void pipenetwork::MatrixAssembler::assemble_jacobian() {
...
-    }
-    // construct jacD part
-    update.emplace_back(nnode_ + index_pipe, index_node1, 1);
-    update.emplace_back(nnode_ + index_pipe, index_node2, -1);
-    // construct jacF part (Hazen-Williams)
-    double deriv_pipe_discharge = pipe.second->deriv_hazen_williams_discharge();
-    update.emplace_back(nnode_ + index_pipe, nnode_ + index_pipe,
-                        deriv_pipe_discharge);
+  // case 1
+  if (pressure <= node->min_pressure()) {
```
Comment (kks32) [inline src/matrix_assembler.cc]: Don't you need a tolerance for <=?

File: src/mesh.cc [hunk 16]
```diff
@@ -1,3 +1,6 @@
+
+#include <mesh.h>
```
Comment (kks32) [inline src/mesh.cc]: remove this

File: src/node.cc [hunk 17]
```diff
@@ -0,0 +1,40 @@
+#include "node.h"
+#include <Eigen/Dense>
```
Comment (kks32) [inline src/node.cc]: Include headers in `node.h`

File: src/node.cc [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+#include <cmath>
+
+Eigen::VectorXd pipenetwork::Node::compute_poly_coefficients(
```
Comment (kks32) [inline src/node.cc:6]: Can you pass three arrays instead? Could you write the formula, so I can see what you are doing here?

File: src/node.cc [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+  ret << a, b, c, d;
+  return ret;
+}
+
+Eigen::VectorXd pipenetwork::Node::get_pdd_poly_coef_1() {
+  double x1 = minimum_pressure_;
+  double f1 = 0;
+  double x2 = minimum_pressure_ + pdd_smoothing_delta_;
+  double f2 = std::pow(
+      (x2 - minimum_pressure_) / (normal_pressure_ - minimum_pressure_), 0.5);
```
Comment (kks32) [inline src/node.cc:27]: x2 - min_pressure is just `pdd_smoothing_delta_`? no?

**Review by kks32 — CHANGES_REQUESTED** [review]
This looks great. Please also run through clang-format and remove any code commented out if you no longer need them.

File: include/matrix_assembler.h (hunk 3 above)
Comment (rewu1993) [reply include/matrix_assembler.h]: If you mean update then yes.

File: include/node.h (hunk 4 above)
Comment (rewu1993) [reply include/node.h]: I dont think we need ishead() anymore. I will remove it

File: include/node.h (hunk 6 above)
Comment (rewu1993) [reply include/node.h]: I thought maybe people want to make different nodes to have difference min/norm pressures

File: src/eigen_gmres.cc (hunk 7 above)
Comment (rewu1993) [reply src/eigen_gmres.cc:34]: Yes! No need for constraints since it is included inside the jacobian.

File: include/node.h (hunk 4 above)
Comment (kks32) [reply include/node.h]: OK

File: src/matrix_assembler.cc [hunk 20]
```diff
@@ -31,111 +35,55 @@ void pipenetwork::MatrixAssembler::assemble_variable_vector() {
...
-  }
-
-  // Iterate through all nodes
-  for (const auto& node : global_nodes_) {
-    Index index = node.first;
-    // Calculate the nodal balance residual (demand part)
-    if (node.second->isdischarge())
-      residual_vec_->coeffRef(index) -= (-1) * node.second->discharge();
+// assember the pressure demand part of jacobian matrix
+void pipenetwork::MatrixAssembler::construct_demand_jac(
```
Comment (kks32) [inline src/matrix_assembler.cc]: Can we not return a vector instead of passing it as a reference? if we are going to modify it?

**Review by kks32 — CHANGES_REQUESTED** [review]

File: src/node.cc (hunk 18 above)
Comment (rewu1993) [reply src/node.cc:6]: I simply did what wntr does for this part. I guess they just try to use polynomial to approximate the transition stage for of the piecewise function such as the slope behavior of (x)^-(0.5) as x approaches 0. I will write the formula though

File: src/matrix_assembler.cc (hunk 20 above)
Comment (rewu1993) [reply src/matrix_assembler.cc]: What you mean about returning a vector?

File: include/matrix_assembler.h (hunk 3 above)
Comment (kks32) [reply include/matrix_assembler.h]: Can we make it return a vector of update instead?

File: include/node.h (hunk 6 above)
Comment (kks32) [reply include/node.h]: I'm happy with setting pressures, but having all variables with getters/setters means it breaks the philosophy of having classes

Comment (kks32) [reply include/node.h]: Could we instead have a JSON or something format to read all geometrical and pressure properties of a pipe?

Comment (kks32) [reply include/node.h]: if not maybe remove all unused getters

File: src/matrix_assembler.cc (hunk 20 above)
Comment (kks32) [reply src/matrix_assembler.cc]: instead of void could we return?

File: src/node.cc (hunk 18 above)
Comment (kks32) [reply src/node.cc:6]: OK

**Review by kks32 — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 21]
```diff
@@ -53,6 +53,7 @@ SET(pipe_network_src
   src/matrix_assembler.cc
   src/mesh.cc
   src/pipe.cc
+        src/node.cc
```
Comment (kks32) [inline CMakeLists.txt:56]: Could you format these please

File: tests/todini_example_test.cc [hunk 22]
```diff
@@ -46,75 +47,155 @@ TEST_CASE("Example network is tested", "[NR method]") {
...
+//                  << std::endl
+//                  << "residual = " << std::endl
+//                  << std::endl
+//                  << (*residual_vec) << std::endl
+//                  << "variable = " << std::endl
+//                  << (*variable_vec) << std::endl
+//                  << std::endl;
+
+      unsigned nr_iter = 1;
+      while (nr_iter < 30) {
```
Comment (kks32) [inline tests/todini_example_test.cc]: Does it converge in 30 iterations?

Comment (kks32) [reply tests/todini_example_test.cc]: why 30?

Comment (rewu1993) [reply tests/todini_example_test.cc]: It is just an arbitrary number for testing purpose. Usually it converges in 10+.

File: src/matrix_assembler.cc (hunk 20 above)
Comment (rewu1993) [reply src/matrix_assembler.cc]: I am trying to construct the sparse matrix. Is it easier to maintain a single triplets vector here? Or you mean return a vector then append this vector to original one? Whats the difference?

File: tests/todini_no_demand.cc [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+              << std::endl
+              << "residual = " << std::endl
+              << std::endl
+              << (*residual_vec) << std::endl
+              << "variable = " << std::endl
+              << (*variable_vec) << std::endl
+              << std::endl;
+
+    unsigned nr_iter = 1;
+    while (nr_iter < 20) {
```
Comment (kks32) [inline tests/todini_no_demand.cc]: Could you please write a for loop instead? and please use ++nr_iter instead of nr_iter++, pre-increment is faster

**Review by kks32 — CHANGES_REQUESTED** [review]
for instead of while in todini test

**Review by kks32 — APPROVED** [review]

**Pull request merged**